lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+                Architecture of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("arch", None)
+
+    @property
+    def epoch(self) -> Optional[str]:
+        """Get epoch of package.
+
+        Returns:
+            (Optional[str]):
+                Epoch of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("epoch", None)
+
+    @property
+    def version(self) -> Optional[str]:
+        """Get version of package.
+
+        Returns:
+            (Optional[str]):
+                Version number of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("version", None)
+
+    @property
+    def fullversion(self) -> Optional[str]:
+        """Get full version of package.
+
+        Returns:
+            (Optional[str]):
+                Full version of package. Defaults to None if package is absent.
+        """
+        if self.absent:
+            return None
+
+        return (
+            f"{(f'{self.epoch}:' if self.epoch else '')}"
+            f"{self.version}{(f'-{self.release}' if self.release else '')}"
+        )
+
+    @property
+    def release(self) -> Optional[str]:
+        """Get release of package.
+
+        Returns:
+            (Optional[str]):
+                Release of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from.
+
+        Returns:
+            (Optional[str]):
+                Repository of package: Defaults to None if package is absent.
+        """
+        return self.__data.get("repo", None)
+
+
+def _load_cache(
+    cache_type: Literal["installed", "available"]
+) -> List[Dict[str, Union[str, _PackageState]]]:
+    """Load DNF cache information.
+
+    Args:
+        cache_type (Literal["installed", "available"]):
+            Cache information category to load.
+
+    Raises:
+        DNFExecutionError: Raised if incorrect cache type is passed.
+
+    Returns:
+        (List[Dict[str, Union[str, _PackageState]]]):
+            Loaded cache information.
+    """
+    if cache_type not in ["installed", "available"]:
+        raise DNFExecutionError(
+            f"Cache type must be either installed or available, not {cache_type}"
+        )
+
+    result = []
+    state = {"installed": _PackageState.INSTALLED, "available": _PackageState.AVAILABLE}
+    for pkg_name, pkg_version, pkg_repo in itertools.filterfalse(
+        lambda x: len(x) != 3,
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
+            )
+        ],
+    ):
+        placeholder = pkg_name.split(".")
+        epoch = None  # Not every DNF package has an epoch.
+        charset, buffer = deque(itertools.chain(pkg_version)), []
```

Comment:
You are right; I can get away with just `deque(pkg_version)`. I just threw `itertools.chain(...)` here because I was like "iterable - I need this to be iterable" without really thinking that wait "str is iterable too!" :smile:

Reply:
This style of parsing the package version string was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb